processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -1197,6 +909,39 @@ long getTimestamp()
     }
 
     @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      TimeAndDims that = (TimeAndDims) o;
+
+      if (timestamp != that.timestamp) {
+        return false;
+      }
+      for (int i = 0; i < dims.length; i++) {
```

Comment:
I think the type comparison in TimeAndDimsComp is actually unnecessary (the dims will have the same types up to numComparisons since the same dimension ordering is maintained), I shouldn't have added that in.

I guess the current equals/hashcode() in this PR doesn't trigger an exception since it's being used only by "no sort" IncrementalIndex during GroupBy merging where the dimension count will be the same across rows?

I think it'd be fine to reuse TimeAndDimsComp as @gianm suggested or to add a similar dims length check to equals(), if you feel like avoiding the name lookups.